client/main.lua

```diff
@@ -196,36 +196,32 @@ end
 local function openVehCatsMenu(category, targetVehicle)
     local vehMenu = {}
 
-    for k, v in pairs(VEHICLES) do
-        if VEHICLES[k].category == category then
-            if not config.vehicles[k] then
-                lib.print.debug('Vehicle not found in config.vehicles. Skipping: '..k)
-            elseif type(config.vehicles[k].shop) == 'table' then
-                for _, shop in pairs(config.vehicles[k].shop) do
-                    if shop == insideShop then
-                        vehMenu[#vehMenu + 1] = {
-                            title = ('%s %s'):format(v.brand, v.name),
-                            description = locale('menus.veh_price')..lib.math.groupdigits(v.price),
-                            serverEvent = 'qbx_vehicleshop:server:swapVehicle',
-                            args = {
-                                toVehicle = v.model,
-                                targetVehicle = targetVehicle,
-                                closestShop = insideShop
-                            }
-                        }
+    local function insertVehicle(data)
+        vehMenu[#vehMenu + 1] = {
+            title = ('%s %s'):format(data.brand, data.name),
+            description = locale('menus.veh_price')..lib.math.groupdigits(data.price),
+            serverEvent = 'qbx_vehicleshop:server:swapVehicle',
+            args = {
+                toVehicle = data.model,
+                targetVehicle = targetVehicle,
+                closestShop = insideShop
+            }
+        }
+    end
+
+    for k, vehicle in pairs(VEHICLES) do
+        if vehicle.category == category then
+            local vehicleShop = config.models[k] or config.categories[vehicle.category] or config.default
+            if not vehicleShop then
+                lib.print.debug('Vehicle not found in config. Skipping: '..k)
+            elseif type(vehicleShop) == 'table' then
+                for i = 1, #vehicleShop do
+                    if vehicleShop[i] == insideShop then
+                        insertVehicle(vehicle)
                     end
                 end
-            elseif config.vehicles[k].shop == insideShop then
-                vehMenu[#vehMenu + 1] = {
-                    title = ('%s %s'):format(v.brand, v.name),
-                    description = locale('menus.veh_price')..lib.math.groupdigits(v.price),
-                    serverEvent = 'qbx_vehicleshop:server:swapVehicle',
-                    args = {
-                        toVehicle = v.model,
-                        targetVehicle = targetVehicle,
-                        closestShop = insideShop
-                    }
-                }
+            elseif vehicleShop == insideShop then
+                insertVehicle(vehicle)
```

Comment:
Since this menu is based on config which is static, we can pre-compute this table for the shop menus on client startup and store in a variable to avoid recomputing every time one opens the menu.

Reply:
pushed a refactor, now it will load when the player opens a category and remain saved until they leave the shop

Reply:
also using a nested function will be the best approach, as we won’t need to pass the table we want to insert to as an argument every time in the loop.

Reply:
I'm suggesting that we just compute this value once and store it on resource start.

Reply:
category and shop changes so we need computing for every category and every shop player in?

Reply:
Yes computing for all vehicles. Since the config is organized in such a way that it makes sense to do one pass on client load.

Reply:
should be good now
https://github.com/Qbox-project/qbx_vehicleshop/pull/95/commits/f4be37eaab02ec502865cbb5fbc2af4272ebe990